Documentation: move Trusty documentation to doc/

gvancuts opened the pull request:
Move the existing Trusty document to the doc/ folder (where
it belongs) and convert the text to ReST.

The Documentation/ folder under hypervisor/ is removed as all
documents should be put under doc/.

All technical information has been preserved or was already
available in other documents.

Fix #150 

Signed-off-by: Geoffroy Van Cutsem <geoffroy.vancutsem@intel.com>

NanlinXie commented:
Can one of the admins verify this patch?

gvancuts commented:
@YadongQi @dongyaozu @jren1 , I shouldn't have messed any technical info during the conversion but I would appreciate if you do a sanity check nevertheless. Thanks in advance!

dbkinder commented on doc/developer-guides/trusty.rst:
```diff
@@ -0,0 +1,186 @@
+Introduction
+************
+
+`Trusty`_ is a set of software components supporting a Trusted Execution Environment (TEE).
```
There's a (bit) assumption here that the reader knows what a TEE is.  Feels like a summary of TEE would be good here,  and a pointer to where folks could read more about it, maybe here: https://datatracker.ietf.org/wg/teep/about/  or here: https://en.wikipedia.org/wiki/Trusted_execution_environment

gvancuts replied:
I'll add one

dbkinder commented on doc/developer-guides/trusty.rst:
```diff
@@ -0,0 +1,186 @@
+   trusted applications executed within the Trusty OS using the kernel drivers
+
+LK (`Little Kernel`_) is a tiny operating system suited for small embedded devices, bootloaders,
+and other environments where OS primitives like threads, mutexes, and timers are needed, but
```
change ``like`` to ``such as``

gvancuts replied:
OK

dbkinder commented on doc/developer-guides/trusty.rst:
```diff
@@ -0,0 +1,186 @@
+
+Trusty Architectural diagram::
+
+                      +---------------------------+
```
This text drawing (and the big one below) are OK for now, but we should convert these to real drawings/images.

gvancuts replied:
I agree, I did not have enough time on my plate today for that but it would be a good enhancement.

dbkinder commented on doc/developer-guides/trusty.rst:
```diff
@@ -0,0 +1,186 @@
+    +-------------------------------------------+
+
+.. note::
+   Trusty OS is running in Secure World in the architecture above.
```
change the sentence ending to:

    ... in the architecture drawing above.

gvancuts replied:
OK

dbkinder commented on doc/developer-guides/trusty.rst:
```diff
@@ -0,0 +1,186 @@
+   * Hypervisor needs to save current world vCPU contexts, and load next world vCPU contexts
+   * Update rdi, rsi, rdx, rbx to next world vCPU contexts
+
+APIs
```
We need to add doxygen API comments to ``hypervisor/include/arch/x86/trusty.h`` and generate the API material properly.

gvancuts replied:
I agree

dbkinder commented on doc/developer-guides/trusty.rst:
```diff
@@ -0,0 +1,186 @@
+   * Hypervisor needs to save current vCPU contexts (Normal World)
+#. ``HC_WORLD_SWITCH``
+
+   * Simulate ARM SMC (Secure Monitor Call, or SMC) instruction to do world switch
```
Why are we talking about ARM instructions?  Is Trusty ARM-specific and we've ported it to run on Intel processors?

gvancuts replied:
No idea :)

@YadongQi @dongyaozu @jren1 , can you enlighten us?

dbkinder commented on doc/developer-guides/trusty.rst:
```diff
@@ -0,0 +1,186 @@
+
+By design, the User OS bootloader (``UOS_Loader``) will trigger the Trusty boot process. The complete bootflow is illustrated below.
+
+.. graphviz:: trusty/boot-flow.dot
```
nice :)

gvancuts replied:
I thought you'd like it ;)

dbkinder commented on doc/developer-guides/trusty.rst:
```diff
@@ -0,0 +1,186 @@
+   :align: center
+   :caption: Trusty Boot Flow
+
+Details
```
Change to:

    As shown in the above figure, here are some details about the Trusty boot flow processing:

gvancuts replied:
"above figure" or "figure above"? (asking because it sounds a little odd to my non-native English speaker ears)

dbkinder commented on doc/developer-guides/trusty.rst:
```diff
@@ -0,0 +1,186 @@
+
+1. UOS_Loader
+
+   a. Load and verify trusty image from virtual disk
```
Trusy should be consistently spelled (with an upper case ``T``)

gvancuts replied:
Fixed

dbkinder commented on doc/developer-guides/trusty.rst:
```diff
@@ -0,0 +1,186 @@
+   a. Load and verify trusty image from virtual disk
+   #. Allocate runtime memory for trusty
+   #. Do ELF relocation of trusty image and get entry address
+   #. Call ``HC_INITIALIZE_TRUSTY`` with trusty memory base and entry address
```
``HC_INITIALIZE_TRUSTY`` is defined as a number (in hypervisor/include/acrn_hv_defs.h) , so how do you call it?

gvancuts replied:
I suspect it should actually be ``hcall_initialize_trusty``, @YadongQi @dongyaozu @jren1 @anthonyzxu , can you confirm?

dbkinder commented on doc/developer-guides/trusty.rst:
```diff
@@ -0,0 +1,186 @@
+#. Trusty
+
+   a. Booting up
+   #. Call ``HC_WORLD_SWITCH`` to switch back to Normal World if boot completed
```
Same comment as above, ``HC_WORLD_SWITCH`` is a defined number not a function.

dbkinder commented on doc/developer-guides/trusty.rst:
```diff
@@ -0,0 +1,186 @@
+Design
+======
+
+Put Secure World's GPA to very high position: 511G-512G. The PML4/PDPT for Trusty
```
We should use the abbreviation ``GB`` for Gigabytes rather than just ``G``, and there should be a space between the number and the units, ``511 GB - 512 GB``

gvancuts replied:
OK

dbkinder commented on doc/developer-guides/trusty.rst:
```diff
@@ -0,0 +1,186 @@
+======
+
+Put Secure World's GPA to very high position: 511G-512G. The PML4/PDPT for Trusty
+World are separated from Normal World. PD/PT for low memory (<511G) are shared in
```
``(< 511 GB)``

gvancuts replied:
OK

dbkinder commented on doc/developer-guides/trusty.rst:
```diff
@@ -0,0 +1,186 @@
+
+Put Secure World's GPA to very high position: 511G-512G. The PML4/PDPT for Trusty
+World are separated from Normal World. PD/PT for low memory (<511G) are shared in
+both Trusty World's EPT and Normal World's EPT. PD/PT for high memory (>=511G) are
```
``(>= 511 GB)``

gvancuts replied:
OK

dbkinder commented on doc/developer-guides/trusty.rst:
```diff
@@ -0,0 +1,186 @@
+=======
+
+This design will benefit the EPT changes of Normal World. There are requirement to
+modify Normal World's EPT during runtime such as memory increasing, attribute
```
reads better as:

    such as increasing memory, changing attributes, etc.

gvancuts replied:
OK

dbkinder commented on doc/developer-guides/trusty.rst:
```diff
@@ -0,0 +1,186 @@
+
+::
+
+    ABSTRACT EPT hierarchy for 2 Worlds:
```
Text drawing is OK for now, but we should change this to a real image (graphviz or a picture).

gvancuts replied:
Can you do something like this in `graphviz`?

I'll put this on my To-Do list

dbkinder commented on doc/developer-guides/trusty.rst:
```diff
@@ -0,0 +1,186 @@
+       rebase_offset:	rebase the gpa to offset xxx(511G_OFFSET)
+   */
+
+   int create_secure_world_ept(vm_t *vm, uint64_t gpa, uint64_t size, uint64_t rebase_offset)
```
I can't find a ``create_secure_world_ept`` function in any file currently in the hypervisor source tree.  This should also be a reference to doxygen-generated API material and not just written here in the documentation.

gvancuts commented:
Hi @dbkinder , I have addressed all your comments but there are a couple of things to finalize before it can be merged:
* There are a couple of open questions that need to be answered
* The hypercalls were already marked with doxygen and I found a way to automatically integrate them in the doc. I enhanced the doxygen marking with some of the content that was in this file. I think it's pretty neat like this... let me know what you think!
* I have an issue with the 3rd API (`create_secure_world_ept`), I am trying to use the [doxygenfunction directive](https://breathe.readthedocs.io/en/latest/directives.html#doxygenfunction) to include it automatically but it always throws an error [1] and I can't figure out what I'm missing... could you take a look? You'll notice that I have had to move the function definition in a header file, that's because of the [breathe-implementation-filename-extensions](https://breathe.readthedocs.io/en/latest/directives.html#breathe-implementation-filename-extensions) that makes it skip all `*.c` files.

[1] Error when running `make doc`
```
$ make html 

New errors/warnings found, please fix them:
==============================================

/home/gvancuts/work/acrn/acrn-hypervisor/doc/developer-guides/trusty.rst:158: WARNING: doxygenfunction: Cannot find function "create_secure_world_ept" in doxygen xml output for project "Project ACRN" from directory: doxygen/xml
```
The weird thing here is that if I search in `doxygen/xml`, I **do** find that `create_secure_world_ept` function...

gvancuts commented:
Just found what the problem was... I had forgotten to put a `;` at the end of the function declaration in the header file... d'oh!

gvancuts commented on hypervisor/include/arch/x86/trusty.h:
```diff
@@ -121,4 +127,19 @@ void destroy_secure_world(struct vm *vm);
+ * @param gpa_rebased gpa rebased to offset xxx (511G_OFFSET)
+ *
+ */
+static void create_secure_world_ept(struct vm *vm, uint64_t gpa_orig,
```
@YadongQi @dongyaozu @jren1 @anthonyzxu , please confirm it's OK to put this function declaration in the header file. This is needed in order to be able to add doxygen comments to it and be able to import it directly in the Trusty doc that this PR is creating.

Please review carefully the parameters and their definition as the document I used as the basis was outdated in that regard.

gvancuts commented:
New version pushed.
* Rephrased the working related to the use of the SMC instruction
* Reverted back to the original state where the `create_secure_world_ept` was only declared and defined in the `trusty.c` file. Worked around the limitation when using `doxygenfunction` by using a group (of a single function) instead (`doxygengroup` directive)
   Is it expected that the only API listed in the doc is this one? There are other APIs in `trusty.c` and it makes me wonder whether they shouldn't also be listed

dbkinder commented:
@gvancuts @anthonyzxu  Is this ready to merge then?

gvancuts commented:
I've just updated this PR to add one more commit that changes the text-based drawings for real pictures.

Can you do a sanity check to make sure it represents the same thing?

This PR is ready to be merged as far as I'm concerned :)

YadongQi commented on hypervisor/include/common/hypercall.h:
```diff
@@ -322,9 +322,23 @@ int64_t hcall_setup_sbuf(struct vm *vm, uint64_t param);
- * @brief Switch VCPU state between Normal/Secure World.
+ * @brief Switch vCPU state between Normal/Secure World.
+ *
+ * * Hypervisor uses the Secure Monitor Code (SMC) instruction to do
```
As @bzhu5 suggested, remove this SMC word, I think replace with "Hypervisor uses this hypercall to do ..." is OK.

gvancuts replied:
OK, I have updated the PR (and made additional, minor modifications to the doxygen comments). This is what the latest version looks like:

![image](https://user-images.githubusercontent.com/2192801/42283832-ffe6c472-7faa-11e8-8b77-42c6de1e4c3d.png)

What do you think?

YadongQi replied:
That's great. Thanks very much.
